core/services/relay/evm/read/multieventtype.go

```diff
@@ -0,0 +1,156 @@
+
+import (
+	"context"
+	"fmt"
+	"iter"
+	"reflect"
+	"strconv"
+	"strings"
+
+	"github.com/ethereum/go-ethereum/common"
+
+	commontypes "github.com/smartcontractkit/chainlink-common/pkg/types"
+	"github.com/smartcontractkit/chainlink-common/pkg/types/query"
+	"github.com/smartcontractkit/chainlink-common/pkg/values"
+	"github.com/smartcontractkit/chainlink/v2/core/chains/evm/logpoller"
+)
+
+type EventQuery struct {
+	Filter           query.KeyFilter
+	EventBinding     *EventBinding
+	SequenceDataType any
+	IsValuePtr       bool
+	Address          common.Address
+}
+
+func MultiEventTypeQuery(ctx context.Context, lp logpoller.LogPoller, eventQueries []EventQuery, limitAndSort query.LimitAndSort) (sequences iter.Seq2[string, commontypes.Sequence], err error) {
+	defer func() {
+		if err != nil {
+			if len(eventQueries) > 0 {
+				var calls []Call
+				for _, eq := range eventQueries {
+					calls = append(calls, Call{
+						ContractAddress: eq.Address,
+						ContractName:    eq.EventBinding.contractName,
+						ReadName:        eq.EventBinding.eventName,
+						ReturnVal:       eq.SequenceDataType,
+					})
+				}
+
+				err = newErrorFromCalls(err, calls, "", eventReadType)
+			} else {
+				err = fmt.Errorf("no event queries provided: %w", err)
+			}
+		}
+	}()
+
+	for _, eq := range eventQueries {
+		if err = eq.EventBinding.validateBound(eq.Address); err != nil {
+			return nil, err
+		}
+	}
+
+	allFilterExpressions := make([]query.Expression, 0, len(eventQueries))
+	for _, eq := range eventQueries {
+		var expressions []query.Expression
+
+		defaultExpressions := []query.Expression{
+			logpoller.NewAddressFilter(eq.Address),
+			logpoller.NewEventSigFilter(eq.EventBinding.hash),
+		}
+		expressions = append(expressions, defaultExpressions...)
+
+		remapped, remapErr := eq.EventBinding.remap(eq.Filter)
+		if remapErr != nil {
+			return nil, fmt.Errorf("error remapping filter: %w", err)
+		}
+		expressions = append(expressions, remapped.Expressions...)
+
+		filterExpression := query.And(expressions...)
+
+		allFilterExpressions = append(allFilterExpressions, filterExpression)
+	}
+
+	eventQuery := query.Or(allFilterExpressions...)
+
+	queryName := ""
+	for _, eq := range eventQueries {
+		queryName += eq.EventBinding.contractName + "-" + eq.Address.String() + "-" + eq.EventBinding.eventName + "-"
+	}
+	queryName = strings.TrimSuffix(queryName, "-")
+
+	logs, err := lp.FilteredLogs(ctx, []query.Expression{eventQuery}, limitAndSort, queryName)
+	if err != nil {
+		return nil, wrapInternalErr(err)
+	}
+
+	return decodeMultiEventTypeLogsIntoSequences(ctx, logs, eventQueries)
+}
+
+func decodeMultiEventTypeLogsIntoSequences(ctx context.Context, logs []logpoller.Log, eventQueries []EventQuery) (iter.Seq2[string, commontypes.Sequence], error) {
+	type sequenceWithKey struct {
+		Key      string
+		Sequence commontypes.Sequence
+	}
+	sequenceWithKeys := make([]sequenceWithKey, 0, len(logs))
+	eventSigToEventQuery := map[common.Hash]EventQuery{}
```

Comment:
Since event signature hashes are created from a string that combines the event name and parameters it is not unrealistic to get a collision here. For e.g. if you query the transfer log from USDC and USDS contract they can have same event hashes

Reply:
see comment above on this rgd solution

Reply:
this is done - validation allows only one query contract.eventname combination